File: core/src/test/java/uk/lobsterdoodle/namepicker/selection/SelectionGridUseCaseTest.java
package uk.lobsterdoodle.namepicker.selection;

import org.junit.Before;
import org.junit.Test;

import uk.lobsterdoodle.namepicker.events.EventBus;
import uk.lobsterdoodle.namepicker.storage.KeyValueStore;

import static org.mockito.Matchers.anyLong;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static uk.lobsterdoodle.namepicker.selection.SelectionGridUseCase.CURRENT_GRID_COLUMN_COUNT;

public class SelectionGridUseCaseTest {
    private KeyValueStore.Edit edit;
    private KeyValueStore storage;
    private EventBus bus;
    private SelectionGridUseCase useCase;

    @Before
    public void setUp() {
        bus = mock(EventBus.class);
        storage = mock(KeyValueStore.class);
        edit = mock(KeyValueStore.Edit.class);
        when(storage.edit()).thenReturn(edit);
        when(edit.put(anyString(), anyLong())).thenReturn(edit);
        useCase = new SelectionGridUseCase(bus, storage);
    }

    @Test
    public void register_on_bus() {
        verify(bus).register(useCase);
    }

    @Test
    public void on_GridColumnSelectedEvent_for_one_post_SelectionGridChangedEvent_for_single_column() {
        useCase.on(new GridColumnSelectedEvent(1));
        verify(bus).post(new SelectionGridChangedEvent(1, 2));
    }

    @Test
    public void on_GridColumnSelectedEvent_for_two_post_SelectionGridChangedEvent_for_two_columns() {
        useCase.on(new GridColumnSelectedEvent(2));
        verify(bus).post(new SelectionGridChangedEvent(2, 1));
    }

    @Test
    public void on_GridColumnSelectedEvent_One_write_change_to_storage() {
        useCase.on(new GridColumnSelectedEvent(1));
        verify(edit).put(CURRENT_GRID_COLUMN_COUNT, 1);
        verify(edit).commit();
    }

    @Test
    public void on_GridColumnSelectedEvent_Two_write_change_to_storage() {
        useCase.on(new GridColumnSelectedEvent(2));
        verify(edit).put(CURRENT_GRID_COLUMN_COUNT, 2);
        verify(edit).commit();
    }

    @Test
    public void on_LoadSelectionGridPreference_post_SelectionGridChangedEvent_with_preference() {
        when(storage.getLong(CURRENT_GRID_COLUMN_COUNT, 2)).thenReturn(1L);
        useCase.on(new LoadSelectionGridPreference());
        verify(bus).post(new SelectionGridChangedEvent(1, 2));
    }
}

File: core/src/test/java/uk/lobsterdoodle/namepicker/selection/SelectionGridUseCaseTest.kt
package uk.lobsterdoodle.namepicker.selection

import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.whenever
import org.junit.Before
import org.junit.Test
import uk.lobsterdoodle.namepicker.Testing.Util.anyLong
import uk.lobsterdoodle.namepicker.Testing.Util.anyString

import uk.lobsterdoodle.namepicker.events.EventBus
import uk.lobsterdoodle.namepicker.storage.KeyValueStore

import uk.lobsterdoodle.namepicker.selection.SelectionGridUseCase.Companion.CURRENT_GRID_COLUMN_COUNT

class SelectionGridUseCaseTest {
    private lateinit var edit: KeyValueStore.Edit
    private lateinit var storage: KeyValueStore
    private lateinit var bus: EventBus
    private lateinit var useCase: SelectionGridUseCase

    @Before
    fun setUp() {
        bus = mock()
        storage = mock()
        edit = mock()
        whenever(storage.edit()).thenReturn(edit)
        whenever(edit.put(anyString(), anyLong())).thenReturn(edit)
        useCase = SelectionGridUseCase(bus, storage)
    }

    @Test
    fun register_on_bus() {
        verify(bus).register(useCase)
    }

    @Test
    fun on_GridColumnSelectedEvent_for_one_post_SelectionGridChangedEvent_for_single_column() {
        useCase.on(GridColumnSelectedEvent(1))
        verify(bus).post(SelectionGridChangedEvent(1, 2))
    }

    @Test
    fun on_GridColumnSelectedEvent_for_two_post_SelectionGridChangedEvent_for_two_columns() {
        useCase.on(GridColumnSelectedEvent(2))
        verify(bus).post(SelectionGridChangedEvent(2, 1))
    }

    @Test
    fun on_GridColumnSelectedEvent_One_write_change_to_storage() {
        useCase.on(GridColumnSelectedEvent(1))
        verify(edit).put(CURRENT_GRID_COLUMN_COUNT, 1)
        verify(edit).commit()
    }

    @Test
    fun on_GridColumnSelectedEvent_Two_write_change_to_storage() {
        useCase.on(GridColumnSelectedEvent(2))
        verify(edit).put(CURRENT_GRID_COLUMN_COUNT, 2)
        verify(edit).commit()
    }

    @Test
    fun on_LoadSelectionGridPreference_post_SelectionGridChangedEvent_with_preference() {
        whenever(storage.getLong(CURRENT_GRID_COLUMN_COUNT, 2)).thenReturn(1L)
        useCase.on(LoadSelectionGridPreference)
        verify(bus).post(SelectionGridChangedEvent(1, 2))
    }
}